servicetalk-http-netty/src/main/java/io/servicetalk/http/netty/NettyPipelinedConnection.java

```diff
@@ -0,0 +1,362 @@
+            return;
+        }
+        src.subscribe(subscriber);
+    }
+
+    private final class WriteTask {
+        private final Subscriber<? super Resp> subscriber;
+        private final Publisher<Req> requestPublisher;
+        private final Supplier<FlushStrategy> flushStrategySupplier;
+        private final Supplier<WriteDemandEstimator> writeDemandEstimatorSupplier;
+
+        private WriteTask(final Subscriber<? super Resp> subscriber,
+                          final Publisher<Req> requestPublisher,
+                          final Supplier<FlushStrategy> flushStrategySupplier,
+                          final Supplier<WriteDemandEstimator> writeDemandEstimatorSupplier) {
+            this.subscriber = subscriber;
+            this.requestPublisher = requestPublisher;
+            this.flushStrategySupplier = flushStrategySupplier;
+            this.writeDemandEstimatorSupplier = writeDemandEstimatorSupplier;
+        }
+
+        void run() {
+            final PublisherSource<Resp> src;
+            try {
+                src = toSource(connection.write(requestPublisher, flushStrategySupplier,
+                        writeDemandEstimatorSupplier)
+                        .afterFinally(() -> {
+                            WriteTask nextWriteTask = pollWithLockAcquired(writeQueue, writeQueueLockUpdater);
+                            if (nextWriteTask != null) {
+                                nextWriteTask.run();
+                            }
+                        }).merge(new Publisher<Resp>() {
+                            @Override
+                            protected void handleSubscribe(final Subscriber<? super Resp> rSubscriber) {
+                                final Subscriber<? super Resp> firstReadSubscriber;
+                                try {
+                                    firstReadSubscriber =
+                                            offerTryAcquireLock(readQueue, readQueueLockUpdater, rSubscriber);
+                                } catch (Throwable cause) {
+                                    closeConnection(rSubscriber, cause);
+                                    return;
+                                }
+
+                                tryStartRead(firstReadSubscriber);
+                            }
+                        }));
+            } catch (Throwable cause) {
+                handleWriteSetupError(subscriber, cause);
+                return;
+            }
+            src.subscribe(subscriber);
+        }
+    }
+
+    private void handleWriteSetupError(Subscriber<? super Resp> subscriber, Throwable cause) {
+        try {
+            closeConnection(subscriber, cause);
+        } finally {
+            // the lock has been acquired!
+            do {
+                WriteTask nextWriteTask;
+                while ((nextWriteTask = writeQueue.poll()) != null) {
+                    deliverTerminalFromSource(nextWriteTask.subscriber, cause);
```

Comment:
Can we avoid calling external methods inside `finally`? eg: `deliverTerminalFromSource` can throw if both `subscriber.onSubscribe()`  and `subscriber.onError()` throw. Alternative:

```java
private void handleWriteSetupError(Subscriber<? super Resp> subscriber, Throwable cause) {
    try {
        closeConnection(subscriber, cause);
    } catch (Throwable t) {
        subscriber.onSubscribe(EMPTY_SUBSCRIPTION);
        subscriber.onError(t);
    }
    // the lock has been acquired!
    do {
        WriteTask nextWriteTask;
        while ((nextWriteTask = writeQueue.poll()) != null) {
            deliverTerminalFromSource(nextWriteTask.subscriber, cause);
        }
    } while (!releaseLock(writeQueueLockUpdater, this) && tryAcquireLock(writeQueueLockUpdater, this));
}
```

OR just make sure `closeConnection` does not throw and remove the `catch` close above.

Reply:
`closeConnection` already shouldn't throw, but since there is a lock acquired the cleanup code is put in a finally block.

I don't think `deliverTerminalFromSource` should throw. if it does it may result in duplicate notification from our sources. For example:

```
NeverPublisher#subscribe()
subscriber.onSubscribe(..)
catch(cause)
handleExceptionFromOnSubscribe(..)
subscriber.onError(cause)
catch(cause2)  (in Publisher#handleSubscribe(..))
deliverTerminalFromSource(..)
subscriber.onSubscribe(..) // duplicate call introduced by our error recovery
catch(cause3)
handleExceptionFromOnSubscribe(..)
subscriber.onError(cause3) // duplicate call introduced by our error recovery
```

I'll submit a PR.

Reply:
> closeConnection already shouldn't throw, 

Unless `connection.closeAsync()` itself throws? I think we should generally avoid calling external code in `finally` block, it complicates control-flow and hides exception if any code in `try` and in the `finally` block throws.

Reply:
I'll just remove the `finally` as it is defensive programming and is targeted at programming errors or runtime failures (e.g. OOME). However I don't think the "don't call external code in finally block" is a hard rule despite it potentially complicating control flow. The goal of "finally" is two independent code blocks need to execute regardless if the first throws. This can be particularly useful to guarantee asynchronous control flow is preserved if something unexpected happens. If exceptions aren't expected from either code block (e.g. programming error occurs as opposed to data flow error), I would argue it doesn't really matter which exception is propagated (if it did matter exceptions are no longer "unexpected"), but we achieved the goal of preserving async control flow. Granted this is a defensive, but if it does occur debugging in an async control flow will be challenging.

In this case if `deliverTerminalFromSource` throws then the control flow isn't correct here because we wouldn't propagate the exception to all pending subscribers, and we should use a safer alternative instead (not convinced this is the desired behavior though #1034). If the `queue.poll()` throws (isn't expect to) then we cannot reliably continue to drain the queue (may get another exception on the next poll attempt) so this is considered fatal and we let the exception propagate.

> Unless connection.closeAsync() itself throws?

Yes any method in `closeConnection()` could throw. However the method returns an async source and throwing isn't the expected way to propagate an error in this case, but may happen in extreme situations (e.g. OOME).

Reply:
>> I'll just remove the finally 

👌 thanks

>>  I don't think the "don't call external code in finally block" is a hard rule 

Agreed.